tac-operation-lifecycle/tac-operation-lifecycle-logic/src/lib.rs

```diff
@@ -0,0 +1,631 @@
+    },
+    thread,
+    time::Duration,
+};
+
+use anyhow::Error;
+use client::{
+    models::profiling::{BlockchainType, OperationType, StageType},
+    Client,
+};
+use database::{EntityStatus, OrderDirection, TacDatabase};
+use tac_operation_lifecycle_entity::{interval, operation};
+
+pub mod client;
+pub mod database;
+pub mod settings;
+
+use futures::{
+    stream::{select, select_with_strategy, BoxStream, PollNext},
+    StreamExt,
+};
+use settings::IndexerSettings;
+use tokio::sync::Mutex;
+use tracing::{instrument, Instrument};
+use uuid::Uuid;
+#[derive(Debug, Clone)]
+pub struct Job {
+    pub interval: interval::Model,
+}
+
+#[derive(Debug, Clone)]
+pub struct OperationJob {
+    pub operation: operation::Model,
+}
+
+#[derive(Debug, Clone)]
+pub enum IndexerJob {
+    Realtime,                // to generate realtime intervals
+    Interval(Job),           // to request operation IDs within interval
+    Operation(OperationJob), // to request profiling data for the operations
+}
+
+impl OperationType {
+    pub fn from_str(s: &str) -> Self {
+        serde_json::from_str(&format!("\"{}\"", s)).unwrap_or(OperationType::ErrorType)
+    }
+
+    pub fn to_id(&self) -> i32 {
+        match self {
+            OperationType::Pending => 1,
+            OperationType::TonTacTon => 2,
+            OperationType::TacTon => 3,
+            OperationType::TonTac => 4,
+            OperationType::Rollback => 5,
+            OperationType::Unknown => 6,
+            OperationType::ErrorType => 0,
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        // SCREAMING_SNAKE_CASE
+        serde_json::to_string(self)
+            .unwrap()
+            .trim_matches('"')
+            .to_string()
+    }
+
+    pub fn is_finalized(&self) -> bool {
+        self != &OperationType::Pending && self != &OperationType::Unknown
+    }
+}
+
+impl StageType {
+    pub fn to_id(&self) -> i32 {
+        match self {
+            StageType::CollectedInTAC => 1,
+            StageType::IncludedInTACConsensus => 2,
+            StageType::ExecutedInTAC => 3,
+            StageType::CollectedInTON => 4,
+            StageType::IncludedInTONConsensus => 5,
+            StageType::ExecutedInTON => 6,
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        format!("{:?}", self)
+    }
+}
+
+impl BlockchainType {
+    pub fn to_string(&self) -> String {
+        format!("{:?}", self)
+    }
+}
+
+pub struct Indexer {
+    settings: IndexerSettings,
+    // This boundary timestamp will used for distinguish between historical and realtime intervals
+    // The main difference lays within interval processing
+    //  - the historical intervals are fetched by selected boundaries
+    //  - the realtime intervals are always fetched FROM the realtime boundary
+    //    (to avoid remote service sync issues)
+    // The boundary can be updated on inserting new operations
+    realtime_boundary: AtomicU64,
+    database: Arc<TacDatabase>,
+}
```

Comment:
NIT: consider moving it to sep file, for example `indexer.rs`. it's common practice to use `lib.rs` as place where to define all public interface, like

```
mod ...
mod ...
pub use ...
pub use ...
pub use ...
pub use ...
```

it's ok to put logic inside this file, but only for small libraries. I think it would be better to split this into files

Reply:
Moved in https://github.com/blockscout/blockscout-rs/pull/1325/commits/23f9ce43046c2811437809b5997a36d77ce08fb3